data/transactions/logic/eval.go

```diff
@@ -3512,39 +3624,43 @@ func opSetByte(cx *EvalContext) error {
 	return nil
 }
 
-func opExtractImpl(x []byte, start, length int) ([]byte, error) {
+func extractCarefully(x []byte, start, length uint64) ([]byte, error) {
+	if start > uint64(len(x)) {
+		return nil, fmt.Errorf("extraction start %d beyond length: %d", start, len(x))
+	}
 	end := start + length
-	if start > len(x) || end > len(x) {
-		return nil, errors.New("extract range beyond length of string")
+	if end < start {
+		return nil, fmt.Errorf("extraction end exceeds uint64")
+	}
+	if end > uint64(len(x)) {
+		return nil, fmt.Errorf("extraction end %d beyond length: %d", end, len(x))
```

Comment:
nit: `end x beyond length` sounds as if the number exceeds length by `x` units. What about `end x is beyond length`?